### PR TITLE
Add support to a configuration file

### PR DESCRIPTION
**Pull request by malukenho**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -27,6 +30,24 @@ directory or file to be linted.
 </tr>
 </table>
 
+### Configuration
+
+**Xulieta** tries to find a `xulieta.yaml` file in the root of your project
```
Comment (Ocramius) [inline README.md]: nnnnnnnnngggggggghhhhhhh YAML

Is there at least a JSON-Schema somewhere under the organisation?

File: bin/xulieta [hunk 2]
```diff
@@ -31,7 +32,7 @@ call_user_func(static function (): void {
     require $autoloadPath;
 
     $application = new Application('Xulieta', '1.0.0');
```
Comment (Ocramius) [inline bin/xulieta]: Maybe `ocramius/package-versions` to be used here?

File: bin/xulieta [hunk 3]
```diff
@@ -31,7 +32,7 @@ call_user_func(static function (): void {
...
-    $application->add(new App());
+    $application->add(new App('Xulieta', ConfigFile::loadInDirectory(__DIR__)));
```
Comment (Ocramius) [inline bin/xulieta]: `Xulieta` repeated twice: why does `App` need the name?

File: src/Command/App.php [hunk 4]
```diff
@@ -5,23 +5,33 @@
...
 use function sprintf;
 
 final class App extends Command
 {
     private bool $errorOccurred = false;
 
+    /** @var string[]  */
+    private array $config;
+
+    /** @param string[] $config */
```
Comment (Ocramius) [inline src/Command/App.php]: I'd be more precise here:

 * are these `array<int, string>`
 * are these `@psalm-var list<string>` maybe?
 * are the values of this list `@psalm-var non-empty-string`, resulting in `@psalm-var list<non-empty-string>`? Or `@psalm-var list<non-empty-string, string>`? Maybe the config could use a better-shaped VO?

File: src/Command/App.php [hunk 5]
```diff
@@ -30,7 +40,7 @@ protected function configure() : void
             ->addArgument(
                 'directory',
                 InputArgument::REQUIRED,
-                'Path to find *.rst and *.md files'
+                'Path to find documentation files'
```
Comment (Ocramius) [inline src/Command/App.php]: `Path where to find documentation files`

File: src/Command/App.php [hunk 6]
```diff
@@ -43,18 +53,20 @@ protected function configure() : void
     protected function execute(InputInterface $input, OutputInterface $output) : int
     {
         $directory = $input->getArgument('directory');
-        $finder    = (new DocFinder($directory))->__invoke();
 
-        $output->writeln("\nFinding documentation files on <info>" . $directory . "</info>\n");
+        $documentFormatHandler = new MultipleDocumentationFormat(...array_map(
+            static fn (string $class) => new $class(),
+            $this->config['plugins']
```
Comment (Ocramius) [inline src/Command/App.php]: Oh, so `$this->config` is a `@psalm-var array{plugins: list<class-string<PluginInterfaceThing>>}` ;-)

File: src/Command/App.php [hunk 7]
```diff
@@ -43,18 +53,20 @@ protected function configure() : void
...
+        ));
 
-        $documentFormatHandler = new MultipleDocumentationFormat(
-            new RstDocumentationFormat(),
-            new MarkdownDocumentationFormat()
-        );
+        $finder = (new DocFinder($directory, $documentFormatHandler->supportedExtensions()))
+            ->__invoke($this->config['exclude_dirs']);
```
Comment (Ocramius) [inline src/Command/App.php]: Similar to the above, the config is also an `@psalm-var array{exclude_dirs: list<non-empty-string>}`

File: src/Config/ConfigFile.php [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+                $dir . '/../../xulieta.yaml',
+                $dir . '/../xulieta.yaml',
+                $dir . '/xulieta.yaml',
+            ],
+            'is_file'
+        ));
+
+        $config = [];
+        if ($configFiles !== false) {
+            $config = (array) Yaml::parse(file_get_contents(realpath($configFiles)));
```
Comment (Ocramius) [inline src/Config/ConfigFile.php]: I would suggest adding JSON-Schema validation at this path, before calling the processor.

File: src/Config/ConfigFile.php [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+use function realpath;
+
+final class ConfigFile
+{
+    /**
+     * @return string[]
+     */
+    public static function loadInDirectory(string $dir) : array
+    {
+        $configFiles = current(array_filter(
```
Comment (Ocramius) [inline src/Config/ConfigFile.php]: Wouldn't `current` default to the last entry of this array? Or would it pick the first? What does mutation testing tell us here?

File: src/DocFinder.php [hunk 10]
```diff
@@ -5,17 +5,24 @@
...
 use function dirname;
 use function is_dir;
+use function sprintf;
 
 final class DocFinder
 {
     private string $directoryOrFile;
 
-    public function __construct(string $directoryOrFile)
+    /** @var string[] */
```
Comment (Ocramius) [inline src/DocFinder.php]: would be more specific here: `@var array<int, string>`, or even better `@psalm-var list<string>`

File: src/Format/MarkdownDocumentationFormat.php [hunk 11]
```diff
@@ -25,9 +26,15 @@ public function __construct(?Markdown $parser = null)
         $this->parser    = $parser ?: new Markdown();
     }
 
-    public function canHandler(SplFileInfo $file) : bool
+    /** @return string[] */
```
Comment (Ocramius) [inline src/Format/MarkdownDocumentationFormat.php]: `@psalm-return list<non-empty-string>` would be best here

File: src/Format/MultipleDocumentationFormat.php [hunk 12]
```diff
@@ -21,10 +23,19 @@ public function __construct(DocumentationFormat ...$documentationFormats)
         $this->documentationFormats = $documentationFormats;
     }
 
-    public function canHandler(SplFileInfo $file) : bool
+    /** @return string[] */
```
Comment (Ocramius) [inline src/Format/MultipleDocumentationFormat.php]: Similar to above review: let's be as precise as we can with types: `@psalm-return list<string>` would be more close to what is being actually exposed

File: README.md (hunk 1 above)
Comment (malukenho) [reply README.md]: I'm actually using `symfony/config` here. 
`ConfigFileValidation` is validating the schema.

Comment (Ocramius) [reply README.md]: `symfony/config` and JSON-Schema are not mutually exclusive: adding a schema makes it possible to auto-complete and pre-validate configs without the need of a runtime.

**Pull request merged**